core/src/main/scala/cromwell/core/callcaching/docker/registryv2/DockerRegistryV2AbstractFlow.scala

```diff
@@ -0,0 +1,285 @@
+    case ParsingResult.Ok(header, errors) =>
+      errors foreach { err => logger.warn(err.formatPretty) }
+      header
+    case ParsingResult.Error(error) => throw new RuntimeException(error.formatPretty)
+  }
+}
+
+/**
+  * Implements logic to build a flow that can retrieve a docker hash 
+  * from a registry server conforming to the docker registry API v2 Specification:
+  * https://docs.docker.com/registry/spec/api/
+  */
+abstract class DockerRegistryV2AbstractFlow(httpClientFlow: HttpDockerFlow)(implicit ec: ExecutionContext, materializer: ActorMaterializer) extends DockerFlow {
+  
+  // Http requests take headers as a scala.collection.immutable.Seq
+  // This implicitly converts type Seq collections to immutable
+  private implicit def toImmutableSeq[T](seq: Seq[T]): ImmutableSeq[T] = ImmutableSeq[T](seq: _*)
+
+  // Wraps the Http flow in a retryable flow to enable auto retries
+  final private val httpFlowWithRetry = new HttpFlowWithRetry[DockerHashContext](httpClientFlow, isRetryable = isRetryable).flow
+
+  final private val tokenFlow = {
+    val responseHandlerFlow = Flow[(HttpResponse, DockerHashContext)].mapAsync(1)(Function.tupled(tokenResponseHandler))
+    requestTransformFlow(buildTokenRequest, responseHandlerFlow)
+  }
+
+  final private val manifestFlow = {
+    val responseHandlerFlow = Flow[(HttpResponse, DockerHashContext)].map(Function.tupled(manifestResponseHandler))
+    requestTransformFlow(Function.tupled(buildManifestRequest _), responseHandlerFlow)
+  }
+  
+  /**
+    * Builds a flow for this docker registry
+    */
+  def buildFlow() = GraphDSL.create() { implicit builder =>
+    import GraphDSL.Implicits._
+    
+    // Decouple the token flow from the manifest flow with ".async" 
+    // this way they can run in parallel from each other 
+    // while still maintaining the final ordering
+    val token = builder.add(tokenFlow.async)
+    val manifest = builder.add(manifestFlow.async)
+    val mergeResponses = builder.add(Merge[(DockerHashResponse, DockerHashContext)](3))
+    
+    token.out0 ~> manifest.in
+                  manifest.out0 ~> mergeResponses.in(0)
+                  manifest.out1 ~> mergeResponses.in(1)
+    token.out1       ~>            mergeResponses.in(2)
+    
+    FlowShape(token.in, mergeResponses.out)
+  }
+
+  /**
+    * Returns true if this flow is able to process this docker image,
+    * false otherwise
+    */
+  def accepts(dockerImageIdentifierWithoutHash: DockerImageIdentifierWithoutHash) = dockerImageIdentifierWithoutHash.host.contains(registryHostName)
+  
+  /* Methods that must to be implemented by a subclass */
+
+  /**
+    * (e.g registry-1.docker.io)
+    */
+  protected def registryHostName: String
+
+  /**
+    * (e.g auth.docker.io)
+    */
+  protected def authorizationServerHostName: String
+
+  /**
+    * Builds the list of headers for the token request
+    */
+  protected def buildTokenRequestHeaders(dockerHashContext: DockerHashContext): Seq[HttpHeader]
```

Comment:
I apologize if the answer is "use your eyes, multiple impls are *already* doing different things, jackass!"